bob-apps/bin/blob_recovery_tool/utils.rs

```diff
@@ -0,0 +1,232 @@
+                    result.push(entry_path.to_string());
+                }
+            }
+            count += 1;
+            if last_log.elapsed() >= Duration::from_secs(1) {
+                log::info!("{} files validated", count);
+                last_log = Instant::now();
+            }
+        }
+    })?;
+    log::info!("Validation completed! {} files validated", count);
+    Ok(result)
+}
+
+pub(crate) fn for_each_file_recursive<F>(path: &Path, function: F) -> AnyResult<()>
+where
+    F: FnMut(&Path),
+{
+    let _ = for_each_file_recursive_inner(path, function)?;
+    Ok(())
+}
+
+fn for_each_file_recursive_inner<F>(path: &Path, mut function: F) -> AnyResult<F>
+where
+    F: FnMut(&Path),
+{
+    for entry in std::fs::read_dir(path)?
+        .into_iter()
+        .filter_map(|res| match res {
+            Ok(entry) => Some(entry),
+            Err(err) => {
+                log::error!("[{:?}] read dir error: {}", path, err);
+                None
+            }
+        })
+    {
+        match entry.file_type() {
+            Ok(file_type) if file_type.is_dir() => {
+                function = for_each_file_recursive_inner(&entry.path(), function)?;
+            }
+            Ok(file_type) if file_type.is_file() => {
+                function(&entry.path());
+            }
+            Err(err) => {
+                log::error!("[{:?}] file type error: {}", path, err);
+            }
+            _ => {}
+        }
+    }
+    Ok(function)
+}
+
+pub(crate) fn validate_index(path: &Path) -> AnyResult<()> {
+    let mut reader = IndexReader::from_path(path)?;
+    let header = reader.read_header()?;
+    reader.read_filter()?;
+    for _ in 0..header.records_count {
+        if reader.is_eof() {
+            return Err(Error::validation_error("invalid records count, can't read more").into());
+        }
+        reader.read_record_header()?;
+    }
+    Ok(())
+}
+
+pub(crate) fn validate_blob(path: &Path) -> AnyResult<()> {
+    let mut reader = BlobReader::from_path(&path)?;
+    reader.read_header()?;
+    while !reader.is_eof() {
+        reader.read_record()?;
+    }
+    Ok(())
+}
+
+pub(crate) fn move_and_recover_blob<P, Q>(
+    invalid_path: P,
+    backup_path: Q,
+    validate_every: usize,
+) -> AnyResult<()>
+where
+    P: AsRef<Path>,
+    Q: AsRef<Path>,
+{
+    std::fs::rename(&invalid_path, &backup_path).with_context(|| "move invalid blob")?;
+
+    log::info!("backup saved to {:?}", backup_path.as_ref());
+    recovery_blob(&backup_path, &invalid_path, validate_every)?;
+    Ok(())
+}
+
+pub(crate) fn recovery_blob<P, Q>(input: &P, output: &Q, validate_every: usize) -> AnyResult<()>
+where
+    P: AsRef<Path>,
+    Q: AsRef<Path>,
+{
+    if input.as_ref() == output.as_ref() {
+        return Err(anyhow::anyhow!(
+            "Recovering into same file is not supported"
+        ));
+    }
+    let should_validate_written = validate_every != 0;
```

Comment:
I think it's better to rename `should_validate_written`, now it's unclear what this flag does

Reply:
Maybe `need_validation` or smth